File: src/main/scala.skip/net/degoes/09-project.scala
/**
 * GRADUATION PROJECT
 *
 * In this section, you will tie together everything you have learned in order to significantly
 * optimize the performance of JVM-based code.
 */
package net.degoes.project

import org.openjdk.jmh.annotations._
import org.openjdk.jmh.infra.Blackhole
import java.util.concurrent.TimeUnit

import zio.Chunk
import scala.util.Random

object dataset1 {
  sealed trait Value
  object Value {
    final case class Text(value: String)    extends Value
    final case class Integer(value: Long)   extends Value
    final case class Decimal(value: Double) extends Value
    case object NA                          extends Value
  }

  final case class Field(name: String)

  final case class Row(map: Map[String, Value]) {
    def apply(field: Field): Value = map(field.name)
  }

  final case class Dataset(rows: Chunk[Row]) { self =>
    def apply(field: Field): Dataset =
      Dataset(
        rows.map(row =>
          if (row.map.contains(field.name)) Row(Map(field.name -> row(field)))
          else Row(Map())
        )
      )

    def *(that: Dataset): Dataset =
      self.binary(that, "*") {
        case (Value.Integer(left), Value.Integer(right)) => Value.Integer(left * right)
        case (Value.Integer(left), Value.Decimal(right)) => Value.Decimal(left * right)
        case (Value.Decimal(left), Value.Decimal(right)) => Value.Decimal(left * right)
        case (Value.Decimal(left), Value.Integer(right)) => Value.Decimal(left * right)
      }

    def +(that: Dataset): Dataset =
      self.binary(that, "+") {
        case (Value.Integer(left), Value.Integer(right)) => Value.Integer(left + right)
        case (Value.Integer(left), Value.Decimal(right)) => Value.Decimal(left + right)
        case (Value.Decimal(left), Value.Decimal(right)) => Value.Decimal(left + right)
        case (Value.Decimal(left), Value.Integer(right)) => Value.Decimal(left + right)
      }

    def -(that: Dataset): Dataset =
      self.binary(that, "-") {
        case (Value.Integer(left), Value.Integer(right)) => Value.Integer(left - right)
        case (Value.Integer(left), Value.Decimal(right)) => Value.Decimal(left - right)
        case (Value.Decimal(left), Value.Decimal(right)) => Value.Decimal(left - right)
        case (Value.Decimal(left), Value.Integer(right)) => Value.Decimal(left - right)
      }

    def /(that: Dataset): Dataset =
      self.binary(that, "/") {
        case (Value.Integer(left), Value.Integer(right)) => Value.Integer(left / right)
        case (Value.Integer(left), Value.Decimal(right)) => Value.Decimal(left / right)
        case (Value.Decimal(left), Value.Decimal(right)) => Value.Decimal(left / right)
        case (Value.Decimal(left), Value.Integer(right)) => Value.Decimal(left / right)
      }

    private def binary(that: Dataset, symbol: String)(
      f: PartialFunction[(Value, Value), Value]
    ): Dataset =
      Dataset(self.rows.zip(that.rows).map { tuple =>
        val (leftRow, rightRow) = tuple

        val map =
          for {
            left                   <- leftRow.map
            (leftName, leftValue)   = left
            right                  <- rightRow.map
            (rightName, rightValue) = right
          } yield s"(leftName ${symbol} rightName)" ->
            ((leftValue, rightValue) match {
              case (left, right) if f.isDefinedAt((left, right)) => f((left, right))
              case (_, _)                                        => Value.NA
            })

        Row(map)
      })
  }
}

/**
 * GRADUATION PROJECT
 *
 * Develop a version of `Dataset` that has a similar API, but which is at least 10x as fast. See how
 * far you can push it (can you get to 100x?).
 *
 * You may assume data is completely homogeneous and that no values are null. However, if ambitious,
 * you may solve the same problem under the assumption of undefined values and heterogeneous data.
 */
@State(Scope.Thread)
@OutputTimeUnit(TimeUnit.SECONDS)
@BenchmarkMode(Array(Mode.Throughput))
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@Threads(16)
class ProjectBenchmark {
  @Param(Array("100", "1000", "10000"))
  var size: Int = _

  object benchmark1 {
    import dataset1._

    var dataset: Dataset = _

    val start: Field  = Field("start")
    val end: Field    = Field("end")
    val netPay: Field = Field("netPay")
  }

  @Setup
  def setupSlow(): Unit = {
    import benchmark1._
    import dataset1._

    val rng: Random = new Random(0L)

    dataset = Dataset(Chunk.fill(size) {
      val start  = rng.between(0, 360)
      val end    = rng.between(start, 360)
      val netPay = rng.between(20000, 60000)

      Row(
        Map(
          "start"  -> Value.Integer(start),
          "end"    -> Value.Integer(end),
          "netPay" -> Value.Integer(netPay)
        )
      )
    })
  }

  @Benchmark
  def baseline(blackhole: Blackhole): Unit = {
    import benchmark1._
    import dataset1._

    val result = (dataset(start) + dataset(end)) / dataset(netPay)

    blackhole.consume(result)
  }

}

File: src/main/kotlin.skip/net/degoes/09-project.kt
/**
 * GRADUATION PROJECT
 *
 * In this section, you will tie together everything you have learned in order to significantly
 * optimize the performance of JVM-based code.
 */
package net.degoes.project

import org.openjdk.jmh.annotations.*
import org.openjdk.jmh.infra.Blackhole
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

import zio.Chunk
import scala.util.Random

object dataset1 {
  interface Value
  final data class Text(val value: String)    : Value
  final data class Integer(val value: Long)   : Value
  final data class Decimal(val value: Double) : Value
  object NA                          : Value

  final data class Field(val name: String)

  final data class Row(val map: Map<String, Value>) {
    operator fun invoke(field: Field): Value = map[field.name]!!
  }

  final data class Dataset(val rows: Chunk<Row>) {
    operator fun invoke(field: Field): Dataset =
      Dataset(
        rows.map { row ->
          if (row.map.contains(field.name)) Row(mapOf(field.name to row(field)))
          else Row(emptyMap())
        }
      )

    operator fun times(that: Dataset): Dataset =
      binary(that, "*", { left, right ->
        when (left) {
          is Integer -> when (right) {
            is Integer -> Integer(left.value * right.value)
            is Decimal -> Decimal(left.value * right.value)
            else       -> NA
          }
          is Decimal -> when (right) {
            is Integer -> Decimal(left.value * right.value)
            is Decimal -> Decimal(left.value * right.value)
            else       -> NA
          }
          else       -> NA
        }
      })

    operator fun plus(that: Dataset): Dataset =
      binary(that, "+", { left, right ->
        when (left) {
          is Integer -> when (right) {
            is Integer -> Integer(left.value + right.value)
            is Decimal -> Decimal(left.value + right.value)
            else       -> NA
          }
          is Decimal -> when (right) {
            is Integer -> Decimal(left.value + right.value)
            is Decimal -> Decimal(left.value + right.value)
            else       -> NA
          }
          else       -> NA
        }
      })

    operator fun minus(that: Dataset): Dataset =
      binary(that, "-", { left, right ->
        when (left) {
          is Integer -> when (right) {
            is Integer -> Integer(left.value - right.value)
            is Decimal -> Decimal(left.value - right.value)
            else       -> NA
          }
          is Decimal -> when (right) {
            is Integer -> Decimal(left.value - right.value)
            is Decimal -> Decimal(left.value - right.value)
            else       -> NA
          }
          else       -> NA
        }
      })

    operator fun div(that: Dataset): Dataset =
      binary(that, "/", { left, right ->
        when (left) {
          is Integer -> when (right) {
            is Integer -> Integer(left.value / right.value)
            is Decimal -> Decimal(left.value / right.value)
            else       -> NA
          }
          is Decimal -> when (right) {
            is Integer -> Decimal(left.value / right.value)
            is Decimal -> Decimal(left.value / right.value)
            else       -> NA
          }
          else       -> NA
        }
      })

    fun binary(that: Dataset, symbol: String, f: (Value, Value) -> Value): Dataset {
      val chunk = (rows.zip(that.rows) as Chunk<scala.Tuple2<Row, Row>>).map { tuple -> Pair(tuple._1, tuple._2) }.map { (leftRow: Row, rightRow: Row) ->
        val newMap: HashMap<String, Value> = HashMap<String, Value>()
        leftRow.map.forEach { leftName: String, leftValue: Value ->
          rightRow.map.forEach { rightName: String, rightValue: Value ->
            val name = "(${leftName} ${symbol} ${rightName})"
            newMap.put(name, f(leftValue, rightValue))
          }
        }
        Row(newMap)
      }

      return Dataset(chunk)
    }
  }
}

/**
 * GRADUATION PROJECT
 *
 * Develop a version of `Dataset` that has a similar API, but which is at least 10x as fast. See how
 * far you can push it (can you get to 100x?).
 *
 * You may assume data is completely homogeneous and that no values are null. However, if ambitious,
 * you may solve the same problem under the assumption of undefined values and heterogeneous data.
 */
@State(Scope.Thread)
@OutputTimeUnit(TimeUnit.SECONDS)
@BenchmarkMode(Mode.Throughput)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@Threads(16)
open class ProjectBenchmark {
  @Param("100", "1000", "10000")
  var size: Int = 0

  object benchmark1 {
    var dataset: dataset1.Dataset? = null

    val start: dataset1.Field  = dataset1.Field("start")
    val end: dataset1.Field    = dataset1.Field("end")
    val netPay: dataset1.Field = dataset1.Field("netPay")
  }

  @Setup
  fun setupSlow(): Unit {
    val rng: Random = Random(0L)

    benchmark1.dataset = dataset1.Dataset(Chunk.fill(size) {
      val start: Long  = rng.between(0, 360).toLong()
      val end: Long    = rng.between(start, 360).toLong()
      val netPay: Long = rng.between(20000, 60000).toLong()

      dataset1.Row(
        mapOf(
          "start"  to dataset1.Integer(start),
          "end"    to dataset1.Integer(end),
          "netPay" to dataset1.Integer(netPay)
        )
      )
    })
  }

  @Benchmark
  fun baseline(blackhole: Blackhole): Unit {
    val ds: dataset1.Dataset = benchmark1.dataset!!
    val result = (ds(benchmark1.start) + ds(benchmark1.end)) / ds(benchmark1.netPay)

    blackhole.consume(result)
  }

}
